game-core/src/main/java/games/strategy/engine/data/changefactory/units/UnitHitsChange.java

```diff
@@ -0,0 +1,57 @@
+package games.strategy.engine.data.changefactory.units;
+
+import games.strategy.engine.data.Change;
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import java.util.Collection;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import org.triplea.java.collections.IntegerMap;
+
+/** A game data change that captures the damage done to a collection of units. */
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class UnitHitsChange extends Change {
+  private static final long serialVersionUID = 2862726651812142713L;
+
+  private final IntegerMap<String> newHits;
+  private final IntegerMap<String> oldHits;
+  private final Collection<String> territoriesToNotify;
+
+  /**
+   * @param hits The amount of hits that the unit should have. It is NOT the difference of the
+   *     existing hits and the new hits.
+   * @param territoriesToNotify The territories that contain all the units
+   */
+  public UnitHitsChange(
+      final IntegerMap<Unit> hits, final Collection<Territory> territoriesToNotify) {
+    this.newHits = new IntegerMap<>();
+    this.oldHits = new IntegerMap<>();
+    hits.entrySet()
+        .forEach(
+            entry -> {
+              this.newHits.add(entry.getKey().getId().toString(), entry.getValue());
+              this.oldHits.add(entry.getKey().getId().toString(), entry.getKey().getHits());
+            });
+    this.territoriesToNotify =
+        territoriesToNotify.stream().map(Territory::getName).collect(Collectors.toList());
+  }
+
+  @Override
+  protected void perform(final GameData data) {
+    for (final String unitId : newHits.keySet()) {
+      final Unit unit = data.getUnits().get(UUID.fromString(unitId));
+      unit.setHits(newHits.getInt(unitId));
+    }
+    for (final String territory : territoriesToNotify) {
+      data.getMap().getTerritory(territory).notifyChanged();
```

Comment:
I don't think it is good idea to allow a save game to run on a new map definition.  I think a lot of things will be FUBAR.

But the only way that I know of figuring out which unit is in which territory is to loop over all the territories and ask if the unit is in them.  This was what the `BombingUnitDamageChange` used to do and I changed that class to be more like this class where the territory is passed in instead of determined during the `perform`.  If you want to have it look up the territories, then I should probably change `BombingUnitDamageChange` to be similar.

Reply:
It kinda seems like we could use a reverse index to be able to efficiently look up the territories of a unit (or just make it a property of the unit even).

It is a design goal to allow maps to change and save games to still work. I don't know how wise that was, it stems from an early design decision that all maps would have just one version installed.

I'll think about this one more, for now let's roll with what you have. Maybe we change it if we make current territory a property of Unit.

Reply:
Currently, if you have a save game for a map and then update the map, the save game will not see the change to the map.  That is because the save game has the entire original map data in it.  You can't play a save game from a map that you don't have, though.  And that's because the images, colors, etc are still read from the map zip/directory.

Thinking about that, if a map is updated with a completely different map picture that includes different connections, territories, etc, then playing the saved game would be really weird.  It would not know of the new connections and it would probably be messed up about the territories.  I'm not sure the save game would even play.

Reply:
IIRC territory names are read from map files for the polygons and knowing where territory boundaries are. Generally agree that it's not good to be playing on altered maps. We have had map makers change territory names and that generally has had bad results for save games and very painful migrations (mainly because we only allow one version of a map at a time and meanwhile we do allow multiple game versions with saves to be running in parallel).